File: src/python/molecule.py
import json
import numpy as np
import re

from basisfunction import BasisFunction
from atom import Atom
from orbital import Orbital

class Molecule:
    def __init__(self):
        self.atoms = list()
        self.basis_functions = list()
        self.orbitals = list()
        self.density_matrix = np.array([])

        self.json_data = None
        self.molden_data = None

        self.x_min = None
        self.x_max = None
        self.y_min = None
        self.y_max = None
        self.z_min = None
        self.z_max = None

    @classmethod
    def from_json(cls, json_file_name):
        instance = cls()

        json_file = open(json_file_name)
        instance.json_data = json.load(json_file)

        instance.read_basis_functions_from_json()
        instance.read_atoms_from_json()
        instance.read_densities_from_json()

        return instance

    @classmethod
    def from_molden(cls, molden_file_name):
        instance = cls()
        print("Loading a molden file ...")

        molden_file = open(molden_file_name)
        instance.molden_data = molden_file.read()
        print("Reading atom data ...")
        instance.read_atoms_from_molden()
        print("Reading GTO data ...")
        instance.read_basis_functions_from_molden()
        print("Reading MO data ...")
        instance.read_orbitals_from_molden()

        print("A molden file has been read!")

        # print(instance.molden_data)

        return instance
        
    def read_basis_functions_from_json(self):
        orbital_labels_list = self.json_data["Molecule"]["MolecularOrbitals"]["OrbitalLabels"]

        coefficients_list = list()
        exponents_list = list()
        shell_list = list()
        positions_list = list()

        for atom in self.json_data["Molecule"]["Atoms"]:
            # print("\tAtom number ", atom["Idx"])
            # print("Coefficients")

            for func in atom["Basis"]:
                # print(func["Coefficients"])
                coefficients_list.append(func["Coefficients"])

                # print("Exponents")
                # print(func["Exponents"])
                exponents_list.append(func["Exponents"])

                # print("Shell")
                # print(func["Shell"])
                shell_list.append(func["Shell"])
                
                positions_list.append(atom["Coords"])

        if not len(coefficients_list) == len(exponents_list) == len(shell_list) == len(positions_list):
            raise Exception("Class Molecule. The coefficients_list, exponents_list, shell_list and positions_list have different shapes")

        for coefficients, exponents, shell, position in zip(coefficients_list, exponents_list, shell_list, positions_list):
            if shell == "s":
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position))
            elif shell == "p":
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="z"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="x"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="y"))
            elif shell == "d":
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="z2"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="xz"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="yz"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="x2y2"))
                self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="xy"))

        if not len(orbital_labels_list) == len(self.basis_functions):
            raise Exception("Class Molecule. len(orbital_labels_list) is not equal len(self.basis_functions)")

        for orbital_label, basis_function in zip(orbital_labels_list, self.basis_functions):
            basis_function.label=orbital_label
            # print(basis_function)

    def read_atoms_from_json(self):
        for atom in self.json_data["Molecule"]["Atoms"]:
            self.atoms.append(Atom(atom["Coords"], atom["ElementLabel"], atom["Idx"], atom["NuclearCharge"]))
        self.update_molecule_limits()

    def read_densities_from_json(self):
        self.density_matrix = np.array(self.json_data["Molecule"]["Densities"]["scfp"])

        if not self.density_matrix.shape[0] == self.density_matrix.shape[1]:
            raise Exception("Class Molecule. Something is wrong with the scfp density")

        if not self.density_matrix.shape[0] == len(self.basis_functions):
            raise Exception("Class Molecule. The number of basis functions is inconsistent with the size of the density matrix")

        # print(np.array2string(self.density_matrix, separator=',', max_line_width=np.inf))
        # print(np.array_equal(self.density_matrix, self.density_matrix.T))

    def read_atoms_from_molden(self):
        # finding the [Atoms] block in the file
        start_block = "[Atoms] AU"
        end_block = "[GTO]"
        start_index = self.molden_data.find(start_block) + len(start_block)
        if start_index == -1:
            raise Exception("Class Molecule. There is no [Atoms] block in the file")

        end_index = self.molden_data.find(end_block, start_index)
        if end_index == -1:
            fragment = self.molden_data[start_index:]
        else:
            fragment = self.molden_data[start_index:end_index].strip()

        # reading atoms info
        atom_lines = fragment.split("\n")
        for line in atom_lines:
            split_line = line.split()
            if not len(split_line) == 6:
                raise Exception("Class Molecule. Unrecognized atom in molden file")

            position = [float(a)*0.529177249 for a in split_line[3:]]  # make a list of coordinates in Angstroms
            self.atoms.append(Atom(position=position, label=split_line[0], number=split_line[1]))

        self.update_molecule_limits()

    def update_molecule_limits(self):
        first_atom_podition = self.atoms[0].position
        
        self.x_min = self.x_max = first_atom_podition[0]
        self.y_min = self.y_max = first_atom_podition[1]
        self.z_min = self.z_max = first_atom_podition[2]

        for atom in self.atoms:
            self.x_min = min(atom.position[0], self.x_min)
            self.x_max = max(atom.position[0], self.x_max)
            self.y_min = min(atom.position[1], self.y_min)
            self.y_max = max(atom.position[1], self.y_max)
            self.z_min = min(atom.position[2], self.z_min)
            self.z_max = max(atom.position[2], self.z_max)


    def read_basis_functions_from_molden(self):
        # finding the [GTO] block in the file
        start_block = "[GTO]"
        end_block = "[5D]"
        start_index = self.molden_data.find(start_block) + len(start_block)
        if start_index == -1:
            raise Exception("Class Molecule. There is no [GTO] block in the file")

        end_index = self.molden_data.find(end_block, start_index)
        if end_index == -1:
            fragment = self.molden_data[start_index:]
        else:
            fragment = self.molden_data[start_index:end_index].strip()

        orbital_shells = ("s", "p", "d", "f")

        # reading gto info
        blocks = fragment.split("\n\n")  # dividing the GTO part into blocks coresponding to atoms
        # if not len(blocks) == len(self.atoms):
        #     raise Exception("Class Molecule. The number of GTO blocks is not equal to the number of atoms")
        for block in blocks:
            for GTO in re.split(r'(?=[spdf])', block)[1:]:  # dividing every block into fragments corresponding to GTOs
                GTO_lines = GTO.split("\n")
                shell = GTO_lines[0].split()[0]
                exponents = [float(a.split()[0]) for a in GTO_lines[1:] if a]
                coefficients = [float(a.split()[1]) for a in GTO_lines[1:] if a]
                number_of_atom = blocks.index(block)
                position = self.atoms[number_of_atom].position

                # print(shell, exponents, coefficients)

                if shell == "s":
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position))
                elif shell == "p":
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="x"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="y"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="z"))
                elif shell == "d":
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="z2"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="xz"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="yz"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="x2y2"))
                    self.basis_functions.append(BasisFunction(coefficients, exponents, shell, position, index="xy"))


            # i = 1
            # while i < len(lines):
            #     label = lines[i].split()[0]
            #     gto_numbers = lines[i].split()[1]
            #     for n in range(0, gto_numbers):
            #         exponent, coeff = lines

    def read_orbitals_from_molden(self):
        start_block = "[MO]"
        start_index = self.molden_data.find(start_block) + len(start_block)
        if start_index == -1:
            raise Exception("Class Molecule. There is no [MO] block in the file")

        fragment = self.molden_data[start_index:]
        blocks = re.split(r'(?=Sym=)', fragment)[1:]
        for block in blocks:
            MO_lines = [l.strip() for l in block.split('\n')]
            # print(MO_lines)
            occ = float(MO_lines[3].split("=")[1])
            # print(occ)
            coefs = [float(a.split()[1]) for a in MO_lines[4:] if a]
            # print(coefs)
            self.orbitals.append(Orbital(occ, coefs))


    def draw_molecule(self, ax):
        for atom in self.atoms:
            ax.scatter(atom.position[0], atom.position[1], atom.position[2], color=atom.ball_color, s=atom.ball_size, alpha=1)

    def scfp_density_at_point(self, point):
        if not self.density_matrix.size == 0:
            basis_functions_values = np.array([bf.value_at_point(point) for bf in self.basis_functions])

            # плотность ρ = φᵀ P φ
            return float(basis_functions_values @ self.density_matrix @ basis_functions_values)
        elif not len(self.orbitals) == 0:
            res = 0
            for orb in self.orbitals:
                if orb.occupancy == 0:
                    continue
                res += orb.occupancy*(self.orbital_value_at_point(point, self.orbitals.index(orb))**2)
            return res

    def orbital_value_at_point(self, point, number):
        basis_functions_values = np.array([bf.value_at_point(point) for bf in self.basis_functions])
        return(np.sum(basis_functions_values @ self.orbitals[number].coefficients))

File: src/cpp/Molecule.cpp
#include "Molecule.h"

#include <fstream>
#include <sstream>
#include <regex>
#include <iostream>
#include <stdexcept>
#include <algorithm>
#include <numeric>
#include <iterator>

Molecule::Molecule()
    : x_min(0), x_max(0), y_min(0), y_max(0), z_min(0), z_max(0), number_of_occupied_orbitals(0) {}

Molecule::Molecule(const std::string& molden_file_name) {
    std::cout << "** Loading a molden file ..." << std::endl;
    molden_data = read_file(molden_file_name);
    
    std::cout << "** Reading atom data ..." << std::endl;
    read_atoms_from_molden();
    #ifdef DEBUG_MODE
    for (auto a : atoms) {
        std::cout << a.repr() << "\n\n";
    }
    std::cout << std::endl;
    #endif

    std::cout << "** Reading GTO data ..." << std::endl;
    read_basis_functions_from_molden();
    #ifdef DEBUG_MODE
    for (auto b : basis_functions) {
        std::cout << b.repr() << "\n\n";
    }
    std::cout << std::endl;
    
    #endif

    std::cout << "** Reading MO data ..." << std::endl;
    read_orbitals_from_molden();
    #ifdef DEBUG_MODE
    for (auto o : orbitals) {
        std::cout << o.repr() << std::endl;
    }
    std::cout << std::endl;
    #endif

    std::cout << "A molden file has been read!" << std::endl;
}

Molecule::~Molecule() {}

std::string Molecule::read_file(const std::string& filename) const {
    std::ifstream file(filename);
    if (!file.is_open())
        throw std::runtime_error("Cannot open molden file: " + filename);

    std::ostringstream ss;
    ss << file.rdbuf();
    return ss.str();
}

void Molecule::read_atoms_from_molden() {
    const std::string start_block = "[Atoms] AU";
    const std::string end_block = "[GTO]";
    size_t start_index = molden_data.find(start_block);
    if (start_index == std::string::npos)
        throw std::runtime_error("Class Molecule. There is no [Atoms] block in the file");

    start_index += start_block.size();
    size_t end_index = molden_data.find(end_block, start_index);
    std::string fragment = (end_index == std::string::npos)
        ? molden_data.substr(start_index)
        : molden_data.substr(start_index, end_index - start_index);

    std::istringstream lines(fragment);
    std::string line;
    while (std::getline(lines, line)) {
        if (line.empty()) continue;

        std::istringstream ls(line);
        std::string label;
        int number, charge;
        double x, y, z;
        std::vector<std::string> tokens;
        std::string token;
        while (ls >> token) tokens.push_back(token);
        if (tokens.size() != 6)
            throw std::runtime_error("Class Molecule. Unrecognized atom line in molden file");

        label = tokens[0];
        number = std::stoi(tokens[1]);
        charge = std::stoi(tokens[2]);
        x = std::stod(tokens[3]) * 0.529177249;
        y = std::stod(tokens[4]) * 0.529177249;
        z = std::stod(tokens[5]) * 0.529177249;

        atoms.emplace_back(std::array<double, 3>{x, y, z}, label, number, charge);
    }

    update_molecule_limits();
}

void Molecule::update_molecule_limits() {
    if (atoms.empty()) return;

    auto first = atoms.front().get_position();
    x_min = x_max = first[0];
    y_min = y_max = first[1];
    z_min = z_max = first[2];

    for (const auto& atom : atoms) {
        const auto& pos = atom.get_position();
        x_min = std::min(x_min, pos[0]);
        x_max = std::max(x_max, pos[0]);
        y_min = std::min(y_min, pos[1]);
        y_max = std::max(y_max, pos[1]);
        z_min = std::min(z_min, pos[2]);
        z_max = std::max(z_max, pos[2]);
    }
}

void Molecule::read_basis_functions_from_molden() {
    const std::string start_block = "[GTO]";
    const std::string end_block = "[5D]";
    size_t start_index = molden_data.find(start_block);
    if (start_index == std::string::npos)
        throw std::runtime_error("Class Molecule. There is no [GTO] block in the file");

    start_index += start_block.size();
    size_t end_index = molden_data.find(end_block, start_index);
    std::string fragment = (end_index == std::string::npos)
        ? molden_data.substr(start_index)
        : molden_data.substr(start_index, end_index - start_index);

    std::regex block_split(R"(\n\s*\n)");
    std::sregex_token_iterator it(fragment.begin(), fragment.end(), block_split, -1);
    std::sregex_token_iterator end;

    std::vector<std::string> blocks(it, end);

    for (size_t block_index = 0; block_index < blocks.size(); ++block_index) {
        const auto& block = blocks[block_index];
        std::regex gto_split(R"((?=[spdf]))");
        std::sregex_token_iterator git(block.begin(), block.end(), gto_split, -1);
        std::vector<std::string> gto_blocks(git, end);

        for (size_t j = 1; j < gto_blocks.size(); ++j) {
            std::istringstream GTO_stream(gto_blocks[j]);
            std::string line;
            std::vector<std::string> lines;
            while (std::getline(GTO_stream, line)) {
                if (!line.empty()) lines.push_back(line);
            }

            if (lines.empty()) continue;
            std::istringstream header(lines[0]);
            std::string shell;
            header >> shell;

            std::vector<double> exponents, coefficients;
            for (size_t i = 1; i < lines.size(); ++i) {
                std::istringstream val_line(lines[i]);
                double exp, coef;
                if (val_line >> exp >> coef) {
                    exponents.push_back(exp);
                    coefficients.push_back(coef);
                }
            }

            auto position = atoms[block_index].get_position();
            if (shell == "s") {
                basis_functions.emplace_back(coefficients, exponents, shell, position);
            } else if (shell == "p") {
                basis_functions.emplace_back(coefficients, exponents, shell, position, "x");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "y");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "z");
            } else if (shell == "d") {
                basis_functions.emplace_back(coefficients, exponents, shell, position, "z2");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "xz");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "yz");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "x2y2");
                basis_functions.emplace_back(coefficients, exponents, shell, position, "xy");
            }
        }
    }
}

void Molecule::read_orbitals_from_molden() {
    std::cout << "begin\n";
    const std::string start_block = "[MO]";

    std::cout << "finding MO block\n";
    size_t start_index = molden_data.find(start_block);
    if (start_index == std::string::npos)
        throw std::runtime_error("Class Molecule. There is no [MO] block in the file");

    std::cout << "creating substring\n";
    start_index += start_block.size();
    std::string fragment = molden_data.substr(start_index);
    // std::cout << fragment << std::endl;

    std::cout << "dividing substring into blocks\n";
    // std::regex mo_split(R"((?=Sym=))");
    // std::sregex_token_iterator it(fragment.begin(), fragment.end(), mo_split, -1);
    // std::sregex_token_iterator end;
    // std::vector<std::string> blocks(it, end);
    // for (auto s : blocks) std::cout << s << std::endl;

    std::string delimiter = "Sym=";
    std::vector<std::string> blocks;

    size_t start = 0;
    size_t end = 0;

    while (end != std::string::npos) {
        end = fragment.find(delimiter, start);
        std::string block = delimiter + fragment.substr(start, end - start);
        // std::cout << block << std::endl;

        blocks.push_back(block);
        start = end + delimiter.length();
    }

    std::cout << "Size of std::vector<std::string> blocks = " << blocks.size() << std::endl;

    std::cout << "reading blocks\n";
    int n = 0;
    for (size_t i = 1; i < blocks.size(); ++i) {
        std::istringstream block(blocks[i]);
        std::string line;
        std::vector<std::string> lines;
        while (std::getline(block, line)) {
            if (!line.empty()) lines.push_back(line);
        }
        if (lines.size() < 5) continue;

        double occ = 0.0;
        {
            std::string occ_line = lines[3];
            size_t pos = occ_line.find("=");
            if (pos != std::string::npos)
                occ = std::stod(occ_line.substr(pos + 1));
        }

        if (occ != 0) {
            n++;
        }

        std::vector<double> coefs;
        for (size_t j = 4; j < lines.size(); ++j) {
            std::istringstream ls(lines[j]);
            double idx, val;
            if (ls >> idx >> val)
                coefs.push_back(val);
        }

        orbitals.emplace_back(occ, coefs);
    }
    number_of_occupied_orbitals = n;

    std::cout << "Size of std::vector<Orbital> orbitals = " << orbitals.size() << std::endl;
}

double Molecule::scfp_density_at_point(const std::array<double, 3>& point) const {
    if (!density_matrix.empty()) {
        std::vector<double> values;
        values.reserve(basis_functions.size());
        for (const auto& bf : basis_functions) {
            auto value = bf.value_at_point(point);
            values.push_back(value);
        }

        // ρ = φᵀ P φ
        double res = 0.0;
        for (size_t i = 0; i < values.size(); ++i) {
            for (size_t j = 0; j < values.size(); ++j) {
                if (i < density_matrix.size() && j < density_matrix[i].size())
                    res += values[i] * density_matrix[i][j] * values[j];
            }
        }
        return res;
    } else if (!orbitals.empty()) {
        double res = 0.0;
        int i = 0;
        for (const auto& orb : orbitals){
            if (orb.get_occupancy() == 0) {
                break;
            }
            double val = orbital_value_at_point(point, i);
            res += orb.get_occupancy() * val * val;
            i++;
        }
        // for (size_t i = 0; i < orbitals.size(); ++i) {
        //     const auto& orb = orbitals[i];
        //     if (orb.get_occupancy() == 0) continue;
        //     double val = orbital_value_at_point(point, i);
        //     res += orb.get_occupancy() * val * val;
        // }
        return res;
    } else {
        return 0.0;
    }
}

double Molecule::orbital_value_at_point(const std::array<double, 3>& point, size_t number) const {
    if (number >= orbitals.size()) return 0.0;

    const auto& coeffs = orbitals[number].get_coefficients();

    if (basis_functions.size() != coeffs.size()) {
        throw std::runtime_error("Class Molecule. The number of basis functions is not equal to the number of coefficients");
    }

    double sum = 0.0;
    auto N = basis_functions.size();
    for (size_t i = 0; i < N; ++i) {
        // if (coeffs[i] == 0) continue;

        sum += basis_functions[i].value_at_point(point) * coeffs[i];
    }

    return sum;
}
